aspnetcore/signalr/security/sample/Startup.cs

```diff
@@ -0,0 +1,108 @@
+﻿#define snippet2 // snippet1
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace SignalR_CORS
+{
+    public class Startup
+    {
+        // This method gets called by the runtime. Use this method to add services to the container.
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+        public void ConfigureServices(IServiceCollection services)
+        {
+        }
+
+#if snippet1
+        #region snippet1
+        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+        {
+            // ... other middleware ...
+
+            // Make sure the CORS middleware is ahead of SignalR.
+            app.UseCors(builder =>
+            {
+                builder.WithOrigins("http://example.com")
+                    .AllowAnyHeader()
+                    .WithMethods("GET", "POST")
+                    .AllowCredentials();
+            });
+
+            // ... other middleware ...
+
+            app.UseSignalR(routes =>
+            {
+                routes.MapHub<ChatHub>("/chatHub");
+            });
+
+            // ... other middleware ...
+        }
+
+        #endregion
+#endif
+#if snippet2
+        #region snippet2
+
+        // In Startup, add a static field listing the allowed Origin values:
+        private static readonly HashSet<string> _allowedOrigins = new HashSet<string>()
+        {
+            // Add allowed origins here. For example:
+            "http://www.mysite.com",
+            "http://mysite.com",
+        };
+
+        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+        {
+            // ... other middleware ...
+
+            // Validate Origin header on WebSocket requests to prevent unexpected cross-site 
+            // WebSocket requests.
+            app.Use((context, next) =>
+            {
+                // Check for a WebSocket request.
+                if (string.Equals(context.Request.Headers["Upgrade"], "websocket"))
+                {
+                    var origin = context.Request.Headers["Origin"];
+
+                    // If there is no origin header, or if the origin header doesn't match 
+                    // an allowed value:
+                    if (string.IsNullOrEmpty(origin) && !_allowedOrigins.Contains(origin))
```

Comment:
```suggestion
                    if (!string.IsNullOrEmpty(origin) && !_allowedOrigins.Contains(origin))
```
@anurse The sample was slightly wrong

Reply:
No, the original content was correct. We want to send a `400` when there is no `Origin` header as well.

Reply:
```suggestion
                    if (string.IsNullOrEmpty(origin) || !_allowedOrigins.Contains(origin))
```

Reply:
If the origin header is null or empty then it could have been the C# or Java client connecting so we should allow them to connect

Reply:
Fair point.